cli/dcoscli/cluster/main.py

```diff
@@ -31,8 +36,20 @@ def main(argv):
 
 @decorate_docopt_usage
 def _main(argv):
+    doc_name = 'cluster'
+
+    # If a cluster is attached and the cluster linker API is reachable,
+    # use cluster_ee.txt.
+    dcos_url = config.get_config_val('core.dcos_url')
+    if dcos_url:
+        try:
+            cluster.get_cluster_links(dcos_url)
```

Comment:
This makes things really slow anytime I want to get the help menu (or run any other) `cluster` subcommand. Is there no other way of determining this information? We should probably be leveraging the cosmos CAPABILITIES API to discover this instead.

E.g.: https://github.com/dcos/cosmos/pull/702/files

Reply:
I'm not too familiar with this API, though I don't get why Cosmos should be responsible for providing this information. The cluster linker is not a cosmos package or service.

For the performance side, we should probably store this kind of information in the config file as it's not meant to change during cluster lifetime. Also this is only requesting when getting the `dcos cluster` help menu, not the main one.

Reply:
Yeah, it's weird that it's part of the cosmos, but that's where it currently lives at the moment (even though it has absolutely nothing to do with cosmos). There has been talk of moving this information somewhere else and building a better schema for it, but it hasn't happened yet.

In the meantime, we should add a CAPABILITY for CLUSTER_LINKING in there and use that here instead of trying to poll the service.

Reply:
That capability would only be for DC/OS EE, I do not see such distinction in the list on the PR you've linked. If I add it there I assume it would also show-up for DC/OS open?

Reply:
Yeah, that seems like a big limitation here. I just don't like that if connecting to an OSS cluster, we are stuck *always* waiting for the 5 (or more) second timeout for any `cluster` command we issue.

If this is the only way at the moment to reliably get the information about linked cluster support, I'd rather do something smarter so that it doesn't take so long on OSS clusters. Do we have a way of reliably detecting OSS vs. EE at the moment? If so, the logic should be something like:
```
if (EE) and (has linked cluster API):
   ...
```

Reply:
Not sure, I saw this issue getting mentioned by @branden on slack (https://jira.mesosphere.com/browse/DCOS_OSS-1160).
Which says : `There's no easy way to determine if a cluster is OSS or EE before logging in.`

Does it mean that there is an easy way once logged in?

Reply:
I think no matter how we check it (it's going to be an HTTP request probably), the performance concern would be addressed by storing some information in the config file, not to have to look-up each time.

Reply:
I disagree. If the linker API is actually there, it should return immediately withotu waiting for the timeout. This amount of time is negligible. It's only when it's not there and we have to suffer through the timeout that it's a problem.

Reply:
>  I just don't like that if connecting to an OSS cluster, we are stuck always waiting for the 5 (or more) second timeout 

This wouldn't hit the request timeout, as the cluster would return early with a 404.

Reply:
I was confused, I thought you were worried about the delay of making an HTTP request each time, that's why I suggested using the config file to cache the info.

Reply:
OK. Makes sense. It's probably fine as is (for now). When we refactor the CLI for 1.12, this sorts of weird edge cases should go away, hopefully.

Reply:
I'm not convinced that this isn't waiting for the timeout for clusters without the linking service:
```
$ time dcos cluster list
                 NAME                                CLUSTER ID                  STATUS    VERSION            URL            
  klueska-05c7d4a0-ee-testing-master    25c036d0-3de4-4b62-94fb-230d6697dc5a   AVAILABLE   1.11-dev  https://35.197.28.151   
  klueska-b093da7d-ee-testing-master    38f071ed-983f-4bc1-bbfc-a6243e0b7eb9  UNAVAILABLE    N/A     https://35.203.152.39   
hybrid-demo-041a3598-ee-testing-master  4efa4abc-90f3-4d7c-b26f-13587a62216a  UNAVAILABLE    N/A      https://35.197.88.36   
     gpu-test-3a3f136a-ee-stable        5000c578-57d4-49d9-8164-10bb27c8a957  UNAVAILABLE    N/A     https://35.185.215.70   
           klueska-l6ds661              62175879-db9e-4187-908b-da4ca25201e1  UNAVAILABLE    N/A      https://52.25.121.75   
     gpu-test-06744c9a-ee-stable        a8866d32-bded-41c9-8dc6-c6a3d52115b3  UNAVAILABLE    N/A     https://35.197.70.255   
      tf-demo-2e2d7cb3-ee-stable        cec885af-748e-4345-b593-a7eca7781c55  UNAVAILABLE    N/A     https://35.197.28.151   
hybrid-demo-d85b8f67-ee-testing-master  e24793ad-1218-4961-90c1-81a2b6dde276  UNAVAILABLE    N/A     https://35.203.167.157  
 eb76b5d5-f32c-426c-b2d7-633e850509c6   eb76b5d5-f32c-426c-b2d7-633e850509c6  UNAVAILABLE    N/A     https://35.197.117.222  

real	1m3.872s
user	0m0.508s
sys	0m0.097s
```

Reply:
I take it back. This is slow because I have so many unreachable clusters. SO it (by definitiion) has to wait for the timeout...

Reply:
would be nice if it were done concurrently

Reply:
Yes, this is slow when a lot of clusters are unavailable.

Doing requests concurrently would be a nice improvement, otherwise at the bare minimum we could at least print each item one-by-one instead of putting all results in memory.